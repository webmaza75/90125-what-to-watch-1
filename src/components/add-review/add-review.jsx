import React, {Fragment, PureComponent} from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import {compose} from 'recompose';
import {NotificationManager} from 'react-notifications';

import ExtendedHeader from '../extended-header/extended-header.jsx';
import {
  itemShape
} from '../../models.js';
import {
  getMovieById
} from '../../reducers/films/selectors.js';
import {TextareaLengths} from '../../consts.js';
import {Operation} from '../../reducers/films/films.js';
import withAddReview from '../../hocs/with-add-review/with-add-review.js';
import {
  historyShape,
  matchShape
} from '../../models.js';

class AddReview extends PureComponent {
  constructor(props) {
    super(props);

    this._handleFormSubmit = this._handleFormSubmit.bind(this);
  }

  render() {
    const {
      movie,
      selectedOption,
      validationError,
      onInputCheck,
      onTextChange,
      isDisabled,
      submiting
    } = this.props;

    if (!movie) {
      return null;
    }

    const {
      title,
      posterImage,
      backgroundImage
    } = movie;

    return <section className="movie-card movie-card--full">
      <div className="movie-card__header">
        <div className="movie-card__bg">
          <img src={backgroundImage} alt={title} />
        </div>

        <h1 className="visually-hidden">WTW</h1>

        <ExtendedHeader movie={movie} />

        <div className="movie-card__poster movie-card__poster--small">
          <img src={posterImage} alt={`${title} poster`} width="218" height="327" />
        </div>
      </div>

      <div className="add-review" onSubmit={this._handleFormSubmit}>
        <form action="#" className="add-review__form">
          {validationError && <div className="sign-in__message">
            <p>{validationError}</p>
          </div>}
          <div className="rating">
            <div className="rating__stars">
              {[1, 2, 3, 4, 5].map((i) => {
                return <Fragment key={`star-${i}`}>
                  <input
                    className="rating__input"
                    id={`star-${i}`}
                    type="radio"
                    name="rating"
                    value={`${i}`}
                    disabled={submiting}
                    checked={selectedOption === `${i}`}
                    onChange={onInputCheck}
                  />
                  <label
                    className="rating__label"
                    htmlFor={`star-${i}`}
                  >
                    {`Rating ${i}`}
                  </label>
                </Fragment>;
              })}
            </div>
          </div>

          <div className="add-review__text">
            <textarea className="add-review__textarea" name="review-text" id="review-text" placeholder="Review text" readOnly={submiting} maxLength={TextareaLengths.MAX_TEXTAREA_LENGHT} onChange={onTextChange
            }></textarea>
            <div className="add-review__submit">
              <button className="add-review__btn" type="submit" disabled={isDisabled}>Post</button>
            </div>

          </div>
        </form>
      </div>
    </section>;
  }

  _handleFormSubmit(event) {
    const {
      onAddComment,
      movie,
      history,
      selectedOption,
      text,
      onFormSubmit,
      onToggleDisabled
    } = this.props;
    onFormSubmit();

    event.preventDefault();
    onAddComment(movie.id, {rating: +selectedOption, comment: text})
      .then(() => {
        onToggleDisabled(true, false);
        history.push(`/film/${movie.id}`);
      })
      .catch((error) => {
        onToggleDisabled(false, false);
        return NotificationManager.error(error.message);
      });
  }
}

AddReview.propTypes = {
  movie: itemShape,
  history: historyShape,
  onAddComment: PropTypes.func,
  text: PropTypes.string,
  selectedOption: PropTypes.string,
  validationError: PropTypes.string,
  isDisabled: PropTypes.bool,
  onInputCheck: PropTypes.func,
  onTextChange: PropTypes.func,
  onFormSubmit: PropTypes.func,
  onToggleDisabled: PropTypes.func,
  submiting: PropTypes.bool,
  match: matchShape
};

const mapStateToProps = (state, {match}) => ({
  movie: getMovieById(state, match.params.id)
});

const mapDispatchToProps = {
  onAddComment: Operation.addComment
};

export {AddReview};

export default compose(
    connect(
        mapStateToProps,
        mapDispatchToProps
    ),
    withAddReview)(AddReview);
