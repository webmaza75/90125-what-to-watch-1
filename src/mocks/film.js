const film = {
  genre: [`Drama`, `Thriller`],
  title: `Aviator`,
  desc: ``,
  picture: `aviator.jpg`,
  year: 2016,
  src: ``
};

export default film;
