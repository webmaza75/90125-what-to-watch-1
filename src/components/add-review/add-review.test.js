import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import {AddReview} from './add-review.jsx';
import film from '../../mocks/film.js';

describe(`AddReview`, () => {
  it(`AddReview correctly renders`, () => {
    const renderer = new ShallowRenderer();
    const result = renderer
      .render(<AddReview
        movie={film}
        text={null}
        validationError={null}
        selectedOption={null}
        submiting={false}
        onCheck={jest.fn()}
        onChangeText={jest.fn()}
        onAddComment={jest.fn()}
        isDisabled={true}
      />);

    expect(result).toMatchSnapshot();
  });
});
