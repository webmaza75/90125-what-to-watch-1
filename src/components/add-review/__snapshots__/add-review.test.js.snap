// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddReview AddReview correctly renders 1`] = `
<section
  className="movie-card movie-card--full"
>
  <div
    className="movie-card__header"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="War of the Worlds"
        src="https://es31-server.appspot.com/wtw/static/film/background/War_of_the_Worlds.jpg"
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <Memo(ConnectFunction)
      movie={
        Object {
          "backgroundColor": "#9B7E61",
          "backgroundImage": "https://es31-server.appspot.com/wtw/static/film/background/War_of_the_Worlds.jpg",
          "description": "As Earth is invaded by alien tripod fighting machines, one family fights for survival.",
          "director": "Steven Spielberg",
          "genre": "Adventure",
          "id": 1,
          "isFavorite": false,
          "picture": "https://es31-server.appspot.com/wtw/static/film/preview/war-of-the-worlds.jpg",
          "posterImage": "https://es31-server.appspot.com/wtw/static/film/poster/War_of_the_Worlds.jpg",
          "rating": 6.5,
          "ratingLevel": "Good",
          "released": 2005,
          "runTime": 116,
          "scoresCount": 386834,
          "src": "https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4",
          "starring": Array [
            "Tom Cruise",
            "Dakota Fanning",
            "Tim Robbins",
          ],
          "title": "War of the Worlds",
          "videoLink": "http://media.xiph.org/mango/tears_of_steel_1080p.webm",
        }
      }
    />
    <div
      className="movie-card__poster movie-card__poster--small"
    >
      <img
        alt="War of the Worlds poster"
        height="327"
        src="https://es31-server.appspot.com/wtw/static/film/poster/War_of_the_Worlds.jpg"
        width="218"
      />
    </div>
  </div>
  <div
    className="add-review"
    onSubmit={[Function]}
  >
    <form
      action="#"
      className="add-review__form"
    >
      <div
        className="rating"
      >
        <div
          className="rating__stars"
        >
          <React.Fragment>
            <input
              checked={false}
              className="rating__input"
              disabled={false}
              id="star-1"
              name="rating"
              onChange={[MockFunction]}
              type="radio"
              value="1"
            />
            <label
              className="rating__label"
              htmlFor="star-1"
            >
              Rating 1
            </label>
          </React.Fragment>
          <React.Fragment>
            <input
              checked={false}
              className="rating__input"
              disabled={false}
              id="star-2"
              name="rating"
              onChange={[MockFunction]}
              type="radio"
              value="2"
            />
            <label
              className="rating__label"
              htmlFor="star-2"
            >
              Rating 2
            </label>
          </React.Fragment>
          <React.Fragment>
            <input
              checked={false}
              className="rating__input"
              disabled={false}
              id="star-3"
              name="rating"
              onChange={[MockFunction]}
              type="radio"
              value="3"
            />
            <label
              className="rating__label"
              htmlFor="star-3"
            >
              Rating 3
            </label>
          </React.Fragment>
          <React.Fragment>
            <input
              checked={false}
              className="rating__input"
              disabled={false}
              id="star-4"
              name="rating"
              onChange={[MockFunction]}
              type="radio"
              value="4"
            />
            <label
              className="rating__label"
              htmlFor="star-4"
            >
              Rating 4
            </label>
          </React.Fragment>
          <React.Fragment>
            <input
              checked={false}
              className="rating__input"
              disabled={false}
              id="star-5"
              name="rating"
              onChange={[MockFunction]}
              type="radio"
              value="5"
            />
            <label
              className="rating__label"
              htmlFor="star-5"
            >
              Rating 5
            </label>
          </React.Fragment>
        </div>
      </div>
      <div
        className="add-review__text"
      >
        <textarea
          className="add-review__textarea"
          id="review-text"
          maxLength={400}
          name="review-text"
          onChange={[MockFunction]}
          placeholder="Review text"
          readOnly={false}
        />
        <div
          className="add-review__submit"
        >
          <button
            className="add-review__btn"
            disabled={true}
            type="submit"
          >
            Post
          </button>
        </div>
      </div>
    </form>
  </div>
</section>
`;
